ntp-proto/src/algorithm/kalman/peer.rs

```diff
@@ -0,0 +1,1077 @@
+};
+
+use super::{
+    config::AlgorithmConfig,
+    matrix::{Matrix, Vector},
+    sqr, PeerSnapshot,
+};
+
+#[derive(Debug, Default, Copy, Clone)]
+struct AveragingBuffer {
+    data: [f64; 8],
+    next_idx: usize,
+}
+
+impl AveragingBuffer {
+    fn mean(&self) -> f64 {
+        self.data.iter().sum::<f64>() / (self.data.len() as f64)
+    }
+
+    fn variance(&self) -> f64 {
+        let mean = self.mean();
+        self.data.iter().map(|v| sqr(v - mean)).sum::<f64>() / ((self.data.len() - 1) as f64)
+    }
+
+    fn update(&mut self, rtt: f64) {
+        self.data[self.next_idx] = rtt;
+        self.next_idx = (self.next_idx + 1) % self.data.len();
+    }
+}
+
+#[derive(Debug, Clone)]
+struct InitialPeerFilter {
+    roundtriptime_stats: AveragingBuffer,
+    init_offset: AveragingBuffer,
+
+    samples: i32,
+}
+
+impl InitialPeerFilter {
+    fn update(&mut self, measurement: Measurement) {
+        self.roundtriptime_stats
+            .update(measurement.delay.to_seconds());
+        self.init_offset.update(measurement.offset.to_seconds());
+        self.samples += 1;
+        debug!(samples = self.samples, "Initial peer update");
+    }
+
+    fn process_offset_steering(&mut self, steer: f64) {
+        for sample in self.init_offset.data.iter_mut() {
+            *sample -= steer;
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+struct PeerFilter {
+    state: Vector,
+    uncertainty: Matrix,
+    clock_wander: f64,
+
+    rtt_stats: AveragingBuffer,
+
+    precision_score: i32,
+    poll_score: i32,
+    desired_poll_interval: PollInterval,
+
+    last_measurement: Measurement,
+    last_packet: NtpPacket<'static>,
+    prev_was_outlier: bool,
+
+    // Last time a packet was processed
+    last_iter: NtpTimestamp,
+    // Current time of the filter state.
+    filter_time: NtpTimestamp,
+}
+
+impl PeerFilter {
+    /// Move the filter forward to reflect the situation at a new, later timestamp
+    fn progress_filtertime(&mut self, time: NtpTimestamp) {
+        debug_assert!(!time.is_before(self.filter_time));
+        if time < self.filter_time {
+            return;
+        }
+
+        // Time step paremeters
+        let delta_t = (time - self.filter_time).to_seconds();
+        let update = Matrix::new(1.0, delta_t, 0.0, 1.0);
+        let process_noise = Matrix::new(
+            self.clock_wander * delta_t * delta_t * delta_t / 3.,
+            self.clock_wander * delta_t * delta_t / 2.,
+            self.clock_wander * delta_t * delta_t / 2.,
+            self.clock_wander * delta_t,
+        );
+
+        // Kalman filter update
+        self.state = update * self.state;
+        self.uncertainty = update * self.uncertainty * update.transpose() + process_noise;
+        self.filter_time = time;
+
+        trace!(?time, "Filter progressed");
+    }
+
+    /// Absorb knowledge from a measurement
+    fn absorb_measurement(&mut self, measurement: Measurement) -> (f64, f64, f64) {
+        // Measurement paramaters
```

Comment:
while we're here: paramaters > parameters

Reply:
Fixed